pyuvsim/simsetup.py

```diff
@@ -285,29 +287,199 @@ def create_mock_catalog(time, arrangement='zenith', array_location=None, Nsrcs=N
     stokes = np.zeros((4, 1, Nsrcs))
     stokes[0, :] = fluxes
     catalog = pyradiosky.SkyModel(names, ra, dec, stokes, 'flat')
-    if return_table:
-        return pyradiosky.skymodel_to_array(catalog), mock_keywords
+    if return_data:
+        catalog = SkyModelData(catalog)
     if get_rank() == 0 and save:
         np.savez('mock_catalog_' + arrangement, ra=ra.rad, dec=dec.rad, alts=alts, azs=azs,
                  fluxes=fluxes)
 
     return catalog, mock_keywords
 
 
+class SkyModelData:
+    """
+    Carries immutable SkyModel data in simple ndarrays.
+
+    This is to facilitate sharing SkyModel objects in MPI, without
+    excessive copying and memory bloat.
+
+    When used with MPI, this can be initialized simultaneously on all
+    processes such that sky_in is provided only on the root process.
+    The data can then be shared across all processes by running the `share`
+    method.
+
+    Parameters
+    ----------
+    sky_in: :class:~`pyradiosky.SkyModel`
+        A valid SkyModel object.
+    """
+
+    spectral_type = None
+    Ncomponents = None
+    ra = None
+    dec = None
+    name = None
+    Nfreqs = None
+    stokes_I = None
+    stokes_Q = None
+    stokes_U = None
+    stokes_V = None
+    freq_array = None
+    reference_frequency = None
+    spectral_index = None
+    polarized = None
+
+    ncomp_attrs = ['stokes_I', 'stokes_Q', 'stokes_U', 'stokes_V',
+                   'ra', 'dec', 'reference_frequency', 'spectral_index', 'name']
```

Comment:
Is this supposed to list attributes that have length `Ncomponents`? I don't think stokes_Q/U/V generally do have that length, do they? I think that they have the length of the number of polarized sources. If that's the case, I'm concerned about where you use it in `__getitem__` to downselect using the component indices. On the other hand, you have a test that looks like it is testing that this works, so I'm confused.

If stokes_Q/U/V do have length `Ncomponents`, I don't see why to use this new object -- it will have the same size as the SkyModel object.

Reply:
Using a new class helps because it keeps the data in numpy arrays with fixed units instead of in Angles/Latitudes/Quantity etc. It's a lot easier to handle with MPI this way.

You're right though, about the stokes_Q/U/V arrays... they don't have length `Ncomponents` unless every source is polarized. This does break __getitem__ so I'll have to fix that too. Thanks for pointing it out!